enhancements/machine-health-checking.md

```diff
@@ -0,0 +1,204 @@
+---
+title: machine-health-checking
+authors:
+  - @bison
+  - @ingvagabund
+  - @enxebre
+  - @cynepco3hahue
+reviewers:
+  - @derekwaynecarr
+  - @michaelgugino
+  - @beekhof
+  - @mrunalp
+approvers:
+  - @enxebre
+  - @bison
+  - @derekwaynecarr
+  - @michaelgugino
+
+creation-date: 2019-09-09
+last-updated: 2019-09-09
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Machine health checking
+
+## Release Signoff Checklist
+
+ - [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs
+
+
+## Summary
+
+Enable opt in for automated health checking and remediation of unhealhty nodes backed by machines.
+
+## Motivation
+
+Reduce administrative overhead to run a cluster and ability to respond to failures of machines.
+
+### Goals
+
+- Providing a default remediation strategy for public clouds - Machine deletion.
+
+- Allowing to define different unhealthy criterias for different pools of machines.
+
+- Allowing new and custom remediation strategies developed outside the machine-healthcheck controller, e.g baremetal reboot.
+
+- Providing administrator-defined short-circuiting of automated remediation.
+
+
+### Non-Goals
+
+- Coordinating individual remediation systems. This must be driven by each system however they see fit, e.g. by annotating affected nodes with state and transition time.
+
+- Recording long-term stable history of all health-check failures or remediations.
```

Comment:
However I expect you to have metrics that can be alerted on and eventually reported back to telemetry - eg whenever we kill reporting a metric, whenever nodes are considered unhealthy reporting a metric.

Reply:
Yes, we definitely want metrics around this. This is noted as a non-goal because there was previously some discussion about creating a CR instance for each remediation attempt, which we want to avoid.